rct229/rules/section5/section5rule21.py

```diff
@@ -0,0 +1,177 @@
+                    if subsurface_classification == DOOR and glazed_area > opaque_area:
+                        fenestration_area_surface += (glazed_area + opaque_area)
+                    else:
+                        fenestration_area_surface += (glazed_area + opaque_area)
+
+            def get_calc_vals(self, context, data=None):
+                above_grade_wall_b = context.baseline
+                above_grade_wall_p = context.proposed
+
+                if (above_grade_wall_b.get["subsurfaces"] ^ above_grade_wall_p.get["subsurfaces"]) or \
+                        (data["total_fenestration_area_b"] == 0.0 ^ data["total_fenestration_area_p"] == 0.0):
+                    # baseline & proposed mismatch:
+                    # baseline has subsurface xor proposed has subsurface or
+                    # baseline fenestration area == 0 xor proposed fenestration area == 0
+                    return {
+                        "total_fenestration_area_surface_b": None,
+                        "total_fenestration_area_b": None,
+                        "total_fenestration_area_surface_p": None,
+                        "total_fenestration_area_p": None
+                    }
+
+                total_fenestration_area_surface_b = 0.0
+                total_fenestration_area_b = data["total_fenestration_area_b"]
+                total_fenestration_area_surface_p = 0.0
+                total_fenestration_area_p = data["total_fenestration_area_p"]
+
+                calc_val = {
+                    "total_fenestration_area_surface_b": total_fenestration_area_surface_b,
+                    "total_fenestration_area_b": total_fenestration_area_b,
+                    "total_fenestration_area_surface_p": total_fenestration_area_surface_p,
+                    "total_fenestration_area_p": total_fenestration_area_p
+                }
+
+                if above_grade_wall_b.get["subsurfaces"] is None:
+                    # wall has no fenestration surfaces
+                    return {
+                        **calc_val
+                    }
+
+                try:
+                    self._helper_calc_val(total_fenestration_area_surface_b, above_grade_wall_b)
+                    self._helper_calc_val(total_fenestration_area_surface_p, above_grade_wall_p)
+                except MissingKeyException as e:
+                    error = f"Missing key in object: {str(e)}, rule check stopped."
+                    return {
+                        "error": error
+                    }
+
+                return {
+                    **calc_val
+                }
+
+            def manual_check_required(self, context, calc_vals=None, data=None):
+                return "error" in calc_vals.keys()
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                if calc_vals["total_fenestration_area_surface_b"] is None and calc_vals["total_fenestration_area_surface_p"] is None \
```

Comment:
I think this can be simplified to:
```
def rule_check(self, context, calc_vals=None, data=None):
  total_fenestration_area_surface_b = calc_vals["total_fenestration_area_surface_b"]
  total_fenestration_area_surface_p = calc_vals["total_fenestration_area_surface_p"]
  total_fenestration_area_b = calc_vals["total_fenestration_area_b"]
  total_fenestration_area_p = calc_vals["total_fenestration_area_p"]

  return (total_fenestration_area_b == ZERO.AREA and total_fenestration_area_p == ZERO.AREA) or
    std_equal(total_fenestration_area_surface_b / total_fenestration_area_b, 
        total_fenestration_area_surface_p / total_fenestration_area_p)

Reply:
Yes, I overthinking the logic in this rule, updated.